src/Draco.Compiler/Internal/Codegen/MetadataCodegen.cs

```diff
@@ -263,17 +249,93 @@ private void EncodeAssembly()
         }
     }
 
+    private void EncodeModule(IModule module, AssemblyReferenceHandle systemRuntime, TypeReferenceHandle systemObject, TypeDefinitionHandle? parentModule = null, int fieldIndex = 1, int procIndex = 1)
```

Comment:
Instead of recursively passing down systemRuntime and systemObject, you could stash them in fields locally. The rest are actually a really smart way to track indices